truffleruby/src/main/java/org/truffleruby/core/binding/BindingNodes.java

```diff
@@ -50,6 +51,36 @@
 @CoreClass("Binding")
 public abstract class BindingNodes {
 
+    @ImportStatic(BindingNodes.class)
+    public abstract static class CreateBindingNode extends RubyBaseNode {
+
+        public abstract DynamicObject execute(MaterializedFrame frame);
+
+        @Specialization(guards = "descriptor == frame.getFrameDescriptor()", limit = "getContext().getOptions().FRAME_VARIABLE_ACCESS_LIMIT")
+        public DynamicObject createBinding(MaterializedFrame frame,
+                @Cached("frame.getFrameDescriptor()") FrameDescriptor descriptor,
+                @Cached("newFrameDescriptor(getContext())") FrameDescriptor bindingDescriptor) {
```

Comment:
Maybe these binding frames should have a different default value (say `null` or a placeholder) so that slots in the descriptor which have this special value in the frame are not considered variables of the binding but just a side-effect of another binding variable being set.

Another idea is to represent Binding variables as a DynamicObject and cache the transformation DynamicObject -> MaterializedFrame. But it's much more involved.